### PR TITLE
custom smtp Service

### PR DESCRIPTION
**Pull request by lc-thomasberger**
refactored so that a custom connection URL (`smtps://user%40gmail.com:pass@smtp.gmail.com/?pool=true`) can be passed as a smtpService config option.

#1447

### CODE REVIEW

**Comment by taylortom** [comment]
Solution looks good to me!

My only slight concern is that we're using the `smtpService` for two completely different uses -- it may be more transparent if we just add a new `smtpConnectionUrl` or something similar.

**Comment by dancgray** [comment]
+1 for a seperate smtpConnectionUrl

File: lib/mailer.js [hunk 1]
```diff
@@ -13,30 +13,46 @@ var Mailer = function () {
...
 
-    if(this.service === '') errors.push('smtpService');
-    if(this.user === '') errors.push('smtpUsername');
-    if(this.pass === '') errors.push('smtpPassword');
-    if(this.from === '') errors.push('fromAddress');
+    if (this.service === '' && !this.useSmtpUrl) {
+      errors.push('smtpService as a well-known Service or a smtpConnectionUrl is required');
+    } else {
+      if (!this.useSmtpUrl) {
+        if (this.service === '') errors.push('smtpService as a well-known Service');
```
Comment (tomgreenfield) [inline lib/mailer.js]: Line 24:
```js
if (this.service === '' && !this.useSmtpUrl) {
```
and lines 27 & 28:
```js
if (!this.useSmtpUrl) {
  if (this.service === '') errors.push('smtpService as a well-known Service');
```
is exactly the same logic?

Comment (lc-thomasberger) [inline lib/mailer.js]: The first one checks if at least one of the two options is defined. If there is, it checks the required values for the well known service.

Comment (taylortom) [inline lib/mailer.js]: @lc-thomasberger I don't see how the second check will ever execute, if both conditions to line 24 are satisfied, it won't ever go into the else.

File: lib/mailer.js [hunk 2]
```diff
@@ -13,30 +13,46 @@ var Mailer = function () {
...
+    smtpConfig = {
+      service: this.service,
+      auth: {
+        user: this.user,
+        pass: this.pass
+      }
+    };
+  }
+
   // Configure the credentials for the specified sevice
```
Comment (taylortom) [inline lib/mailer.js]: Can you change `sevice` -> `service`

File: lib/mailer.js (hunk 1 above)
Comment (taylortom) [inline lib/mailer.js]: I'm finding this block quite hard to read, could we do something like:
```
if(!this.useSmtpUrl) {
        if (this.service === '') errors.push('smtpService as a well-known Service');
        if (this.user === '') errors.push('smtpUsername');
        if (this.pass === '') errors.push('smtpPassword');
        if (this.from === '') errors.push('fromAddress');
    }
```

**Pull request closed without merging**